### PR TITLE
feat: dockerize

### PR DESCRIPTION
**Pull request by maxarndt**
Hey @KohheePeace!

I struggled for a long time using this program in a CD pipeline because of the special dependencies needed by puppeteer.

There are some docker images on the Docker Hub which provide all these dependencies but unfortunately they all have some limitations. Some of them enforce you to call `puppeteer.launch()` with  `{ args: ['--no-sandbox'] }` as argument which is currently not set in the app. Others need to have the `--cap-add=SYS_ADMIN` flag set on `docker run`. The latter is not possible for GitHub Actions for example (see [docs](https://help.github.com/en/actions/building-actions/dockerfile-support-for-github-actions#supported-linux-capabilities)). 

Because of this I decided to dockerize this application to have it working without any other dependencies.
I used the [Dockerfile template](https://github.com/puppeteer/puppeteer/blob/master/docs/troubleshooting.md#running-puppeteer-in-docker) provided in the puppeteer repo and it works quiet well. Because of the `--no-sandbox` limitation of puppeteer I introduced another option (at least for the `from-build` action).

I updated the README.md and uploaded the docker image to my Docker Hub. Please give it a try:
```
docker run --rm -it -v /someDir/my-docusaurus:/workspace maxys/docusaurus-pdf:latest from-build --no-sandbox -o /workspace/build/docs.pdf /workspace/build docs/doc1 myBaseUrl
```
Probably it would be best to link this repo as a Docker Hub image. If we do it this way we have the README.md shown up on the Docker Hub as well.

Now, using this tool in a CD pipeline should work. I will give it a try with GitHub Actions tomorrow. If this runs stable I can create a GitHub Action as well.

What do you think about it?

### CODE REVIEW

**Comment by KohheePeace** [comment]
Cool! thanks for your hard work! I will check this!

**Comment by KohheePeace** [comment]
Actually I'm not a fan of generating PDF during CI/CD for sake of simplicity which is like discussed here (https://github.com/jxnblk/mdx-deck/issues/141) 

Could you add the instructions of the flow to generate PDF and link it automatically for end user can download the PDF ?

**Comment by KohheePeace** [comment]
Is `from-build` not enough ?

In my case I can successfully generate PDF when just pushing github with netlify.

[My flow of using docusaurus]
1. Edit document
2. commit and push the changes to github
3. Netlify automatically run `yarn build` command written in `package.json`

So, in my `package.json`

```json
"scripts": {
    "start": "docusaurus start",
    "build": "docusaurus build && yarn pdf-from-build",
    "pdf-from-build": "npx docusaurus-pdf from-build build/ docs/flutter-crud2a/introduction --output-file build/img/docusaurus.pdf",
    "swizzle": "docusaurus swizzle",
    "deploy": "docusaurus deploy"
  },
```

Important part of scripts.
```json
npx docusaurus-pdf from-build build/ docs/flutter-crud2a/introduction --output-file build/img/docusaurus.pdf
```

So, you can access generated PDF at `https://www.yourdomain.com/img/docusaurus.pdf`

**Comment by maxarndt** [comment]
> Is `from-build` not enough ?
> 
> In my case I can successfully generate PDF when just pushing github with netlify.
> 
> [My flow of using docusaurus]
> 
> 1. Edit document
> 2. commit and push the changes to github
> 3. Netlify automatically run `yarn build` command written in `package.json`
> 
> So, in my `package.json`
> 
> ```json
> "scripts": {
>     "start": "docusaurus start",
>     "build": "docusaurus build && yarn pdf-from-build",
>     "pdf-from-build": "npx docusaurus-pdf from-build build/ docs/flutter-crud2a/introduction --output-file build/img/docusaurus.pdf",
>     "swizzle": "docusaurus swizzle",
>     "deploy": "docusaurus deploy"
>   },
> ```
> 
> Important part of scripts.
> 
> ```json
> npx docusaurus-pdf from-build build/ docs/flutter-crud2a/introduction --output-file build/img/docusaurus.pdf
> ```
> 
> So, you can access generated PDF at `https://www.yourdomain.com/img/docusaurus.pdf`

Sounds great! You should document this somewhere in the docs. I think there are quit a few people out there who are interested to use this with netlify.

It is good to know that running `npx docusaurus-pdf` in the netlify build environment works. 
But it seems this does not work in any node environment because some of the dependencies used in this project depend on other libraries which can not be loaded via npm/npx. 
For example try to run `npx docusaurus-pdf form build ...` command within a `node:10` docker container:
```
docker run -it --rm -v /path/to/your/docusaurus-dir:/workspace --entrypoint=npx node:12 docusaurus-pdf from-build build/ docs/flutter-crud2a/introduction --output-file build/img/docusaurus.pdf
```
This does not work. 
As long as there is no way to execute this tool with `npx` in a regular `node` environment I think there should be a docker image containing this program to ensure this program can be executed everywhere you have at least docker installed.


> Actually I'm not a fan of generating PDF during CI/CD for sake of simplicity which is like discussed here ([jxnblk/mdx-deck#141](https://github.com/jxnblk/mdx-deck/issues/141))

I'm not sure if I understand you correctly. Isn't that what you are doing with your `"build": "docusaurus build && yarn pdf-from-build"`?

> Could you add the instructions of the flow to generate PDF and link it automatically for end user can download the PDF in README?

What do you mean with _"download the PDF in README?"_?

If you just wanted to know the flow of my CD (currently GitHub Actions) pipeline:
1. Checkout repository
2. `npm install`
3. `npm run build`
4. Execute custom GitHub Action which runs dockerized docusaurus-pdf (could be published to [GitHub Action marketplace](https://github.com/marketplace?type=actions) in the future)
5. Upload `build` directory to bucket which is used for static website hosting

As at your site, users can access the PDF at `https://domain.tld/somePath/docs.pdf`.

Basically my pipeline is similar to yours. The only difference is that `npx docusaurus-pdf` fails in my build environment and therefore I use a Docker image which runs `npx docusaurus-pdf`. This ensure all dependencies are available which are needed by `docusaurus-pdf`.

Let me know if anything is unclear or difficult to understand.

**Comment by KohheePeace** [comment]
@maxarndt Thanks for taking your time! 😸 

> The only difference is that npx docusaurus-pdf fails in my build environment and therefore I use a Docker image which runs npx docusaurus-pdf. This ensure all dependencies are available which are needed by docusaurus-pdf.

=> Okay. This is the big point of this PR. 👍 

> > Actually I'm not a fan of generating PDF during CI/CD for sake of simplicity which is like discussed here (jxnblk/mdx-deck#141)
> 
> I'm not sure if I understand you correctly. Isn't that what you are doing with your "build": "docusaurus build && yarn pdf-from-build"?

=> Sorry I just did this for testing. I prefer generating PDF manually. But it is true someone wants to automate this process.

**[Conclusion]**
I understand the problem solved by this PR! So, thanks for your hard work! 💯

**Comment by svict4** [comment]
Hey @maxarndt Nice work on the docker image. I'm wanting to generate the PDF via a Github action as well. Did you have any luck doing so?

**Comment by maxarndt** [comment]
Hey @svict4! Yes, I did create a GitHub action which worked out for me. 
```yaml
# action.yml
name: 'create pdf'
description: 'Create PDf from docusaurus sources'
runs:
  using: 'docker'
  image: 'docker://maxys/docusaurus-pdf:latest'
  args:
    - from-build
    - --no-sandbox
    - -o
    - /github/workspace/build/docs.pdf
    - /github/workspace/build
    - home
    - foo
```

Because there is currently no official docker image of this project I did not published this GitHub action. 
Feel free to use this snippet! And of course feel free to contribute! 🙂
